client/my-sites/checkout/composite-checkout/composite-checkout.tsx

```diff
@@ -700,14 +700,23 @@ export default function CompositeCheckout( {
 				clearSignupDestinationCookie();
 			}
 
+			if ( searchParams.has( 'cancel_to' ) ) {
+				const cancelPath = searchParams.get( 'cancel_to' ) ?? '';
+				// Only allow redirecting to relative paths.
+				if ( cancelPath.match( /^\/[^/]?/ ) ) {
+					page( cancelPath );
+					return;
+				}
+			}
+
 			// Some places that open checkout (eg: purchase page renewals) return the
 			// user there after checkout by putting the previous page's path in the
 			// `redirect_to` query param. When leaving checkout via the close button,
 			// we probably want to return to that location also.
 			if ( searchParams.has( 'redirect_to' ) ) {
 				const redirectPath = searchParams.get( 'redirect_to' ) ?? '';
 				// Only allow redirecting to relative paths.
-				if ( redirectPath.startsWith( '/' ) ) {
+				if ( redirectPath.match( /^\/[^/]?/ ) ) {
```

Comment:
@sirbrillig / @michaeldcain I don't think we're using the composite checkout here but noticed the matching code in masterbar/checkout.tsx. This is to prevent `//example.com` getting treated as safe to redirect.

Can you confirm this code only runs for the "cancel" path? `redirect_to` looks like its handled somewhere else for the "checkout success" path.

Also, how can I test the composite checkout behaves the same / as expected after this change? Is it in use yet? Readme had a note about this still being under devel.

Reply:
In the interest of getting this merged, could we remove the `cancel_to` change for now? When they cancel, the site won't have the correct plan and should be redirected back to the confirm step.

Reply:
> Can you confirm this code only runs for the "cancel" path? redirect_to looks like its handled somewhere else for the "checkout success" path.

That's correct. The post-Checkout handling of `redirect_to` happens [in `getThankYouPageURL()`](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts#L94). I think this code path only runs for the empty cart view that happens when there's an error adding an item to the shopping cart. You can test it by [setting this to `true`](https://github.com/Automattic/wp-calypso/blob/fix%2Fpurchase-meta-alignment/client/my-sites/checkout/composite-checkout/composite-checkout.tsx#L730) and visiting Checkout.

I also tested the similar change to `masterbar/checkout.tsx` and it works as expected, so this LGTM. Thanks for digging deeper to keep them in sync.

> Is it in use yet? Readme had a note about this still being under devel.

Composite Checkout is the only version of Checkout now. Thus, if you buy a Jetpack or WPcom product, you're testing it :). I'll update that README.

Reply:
Excellent, thanks @michaeldcain. It sounds like this could be ready to merge with a bit more testing then.

Just noting that one of the links in your comment was broken for me, but [it seemed to be pointing here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/composite-checkout.tsx#L730), which makes sense